### DISCUSSION TITLE
Designing field-split preconditioners for NS

### DISCUSSION THREAD

**Discussion by lindsayad** [idaholab/moose#24809] [Q&A Modules: Navier-Stokes]
This is meant to be a repository of information for designing field split conditioners for NS. When (or even before) we have options that work robustly for a range of Reynolds numbers, this should go into the MOOSE documentation

My general experience:
- Jacobi for the Schur takes more iterations than something like ASM+(I)LU but it is less expensive so it often wins
- Apparently boomeramg works better on high Re problems when L is large compared to ASM+(I)LU but ASM+(I)LU works better on high Re problems for small mu. This in reference to solving the momentum part of the matrix
- If there is no diagonal, then for Jacobi for Schur, specify `-pc_jacobi_type rowmax`; `-pc_jacobi_type rowsum` appears to be terrible. Note that when using the default `-pc_jacobi_type diagonal` if a 0 is encountered, then 1 is used
- Jacobi doesn't work for the momentum part of the matrix; not strong enough
- For advection dominated, `full` factorization does not compare as poorly to `lower/upper` as for non-advection dominated, in fact it may be faster
- I believe for Aij matrices in PETSc that the block size is always 1. So when doing Schur problems, doing Ainv types of `diag` and `blockdiag` are equivalent and in these cases when using `selfp` Ainv just corresponds to the reciprocal of A's diagonal which for advection dominated leads to Sp poorly approximating S, e.g. even when doing first order upwind with a Reynolds number of 2.2 in our `2d-rc-no-slip.i` channel flow problem, it takes 50-60 linear iterations to solve the Schur complement with `-pc_type lu` for both the outer Schur PC type and the A00 PC type
- `selfp` degrades with increasing Reynolds number as the diagonal of A00 is a poor representation of A00
- `self` with LSC preconditioning appears to be promising for stable discretizations (e.g. Q2-Q1 finite elements). It requires some PETSC development ([see this MR](https://gitlab.com/petsc/petsc/-/merge_requests/6642)) in order to have mesh independent scaling. The current PETSc implementation does not add corrections for stabilized discretizations such as PSPG for finite elements or Rhie-Chow for finite volumes. I'll have to see whether that is required
    - `self` with LSC is indeed a great preconditioner, especially for stable C0 finite elements. See timing in https://github.com/idaholab/moose/issues/24548#issuecomment-1660899338. When doing LSC preconditioning with a KSP solve, it is best to limit the number of outer linear solve iterations, e.g. `full` should be used as the factorization as opposed to `upper,lower,diag`, since the ksp(S,Sp) will be expensive

**Comment by TLWise** [comment]
Would you be able to share a copy of the code that you used to evaluate and determine you findings a presented above? I would like to run my own test with different geometry and variable values with what you have developed.

**Reply by GiudGiud** [reply]
You can use this example: https://github.com/idaholab/moose/blob/next/modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/2d-rc-no-slip.i
we base most of our studies on this one and on a large scale reactor model we can't share.

**Reply by lindsayad** [reply]
I've opened https://github.com/idaholab/moose/pull/24883 if you want to follow along with my field split research. That PR includes a finite element example, which I believe you asked about elsewhere
